source("5. Manuscript code/utils.R")

# The following results file is generated by `Other analyses/RNAscope.analysis.R`
validations <- readRDS("3. Other analyses/data/RNAscope.rds")

#####################################################################################################################
#                                           Figure 4 - RNAscope Validations                                         #
#####################################################################################################################


# ----------------------------------------------------------------------------------------------------------------- #
#                                           Panel A - Mic12/13 pathways vs all                                      #
# ----------------------------------------------------------------------------------------------------------------- #
pathways <- h5read(aggregated.data, file.path(mapping[["microglia"]], "pa")) %>% 
  filter(state %in% c("Mic.12","Mic.13") & Count > 2) %>% 
  mutate(direction.n = if_else(direction == "upregulated", 1, -1),
         signed.pval = -log10(p.adjust) * direction.n) %>% 
  pivot_wider(id_cols = c("Description"), names_from = c(state, state), values_from = c(signed.pval, direction.n)) %>%
  mutate(across(contains("direction.n"), ~replace_na(., 0)),
         tot.direction = direction.n_Mic.12 + direction.n_Mic.13) %>%
  arrange(tot.direction, signed.pval_Mic.12, signed.pval_Mic.13) %>%
  column_to_rownames("Description")

annotate <- c("negative regulation of cell migration",
              "negative regulation of response to oxidative stress",
              "actin binding", "regulation of lipase activity",
              "foam cell differentiation", "lipid catabolic process",
              "lipid catabolic process", "exploration behavior",
              "glial cell migration", "synapse organization",
              "positive regulation of protein polymerization",
              "Platelet activation, signaling and aggregation")
annotate <- which(rownames(pathways) %in% annotate)

pdf(file.path(panel.path, "4A.pdf"), width=embed.width*.7, height=embed.height*1.2)
Heatmap(pathways %>% select(contains("signed.pval")) %>% as.matrix, 
        col = green2purple.less.white(21),
        cluster_rows = F, 
        cluster_columns = F,
        show_row_names = F,
        row_split = pathways$tot.direction,
        column_labels = c("Mic.12", "Mic.13"),
        row_title = " ",
        right_annotation = rowAnnotation(text=anno_mark(at=annotate, labels=rownames(pathways)[annotate])),
        border=T, 
        row_gap = unit(0,"pt"),
        na_col = "grey90",
        height = unit(6,"cm"),
        width = unit(.8, "cm"))
while (!is.null(dev.list()))  dev.off()
rm(pathways, annotate)



# ----------------------------------------------------------------------------------------------------------------- #
#                                           Panel B - Mic12 vs Mic13 pathways                                       #
# ----------------------------------------------------------------------------------------------------------------- #
pathways <- h5read(aggregated.data, file.path(mapping[["microglia"]], "pa.pairwise")) %>% 
  filter(comparison %in% c("Mic.12 vs. Mic.13","Mic.13 vs. Mic.12")) %>%
  dplyr::rename(state=comparison) %>%
  mutate(direction = ifelse(state == "Mic.13 vs. Mic.12", 1, -1), rowname=Description,
         signed.pval = -log10(p.adjust)*direction) %>%
  arrange(signed.pval) %>%
  column_to_rownames("rowname")


annotate = c("MHC class II protein complex", "negative regulation of leukocyte apoptotic process",
             "Antigen processing and presentation", "axonogenesis",
             "extracellular matrix organization", "cell junction assembly", "lamellipodium assembly",
             "lamellipodium assembly and organization", "negative regulation of immune system process",
             "negative regulation of locomotion", "immunological synapse")
annotate <- which(rownames(pathways) %in% annotate)

pdf(file.path(panel.path, "4B.pdf"), width=embed.width*.7, height=embed.height*1.2)
Heatmap(pathways %>% select(signed.pval) %>% as.matrix, 
        col = green2purple(21),
        cluster_rows = F, show_row_names = F, show_column_names = F,
        row_title = " ",
        right_annotation = rowAnnotation(text=anno_mark(at=annotate, labels=rownames(pathways)[annotate])),
        row_split = pathways$direction,
        border=T,
        row_gap = unit(0,"pt"),
        height = unit(6,"cm"),
        width = unit(.4, "cm"))
while (!is.null(dev.list()))  dev.off()
rm(annotate, pathways)



# ----------------------------------------------------------------------------------------------------------------- #
#                                   Panel C - Selective genes differential between Mic12/13/Mac                     #
# ----------------------------------------------------------------------------------------------------------------- #

genes = list(`Mic.12 genes` = c("CD163","HLA-DRA","FOXP1","PELI1","PELI2"),
             `Both`         = c("MSR1","PRKCE"),
             `Mic.13 genes` = c("PTPRG","PPARG","SPP1","TGFBR1","SMAD3","SCIN","WIPF3"),
             `AD risk genes`= c("ADAM10","TREM2","APOE","GPNMB"))

exp <- h5read(aggregated.data, "glia/microglia/gene.exp") %>% filter(gene %in% unlist(genes))

# Summarize "other" microglia (non Mic.12/13/Mac/Mono) gene expression
n.mic <- exp %>% select(state, n) %>% unique %>% pull(n) %>% sum

exp <- exp %>% mutate(collapse.state = !state %in% c("Mic.12","Mic.13","Monocytes","Macrophages"),
                      state.prop = if_else(collapse.state, n/n.mic, 1),
                      state = if_else(collapse.state, "Other", state)) %>% 
  mutate(mean.exp = mean.exp*state.prop,
         pct.exp = pct.exp*state.prop) %>%
  group_by(state, gene) %>% 
  summarise(across(c(mean.exp, pct.exp), sum))


hm <- gheatmap(df = exp, color.by = "mean.exp", size.by="pct.exp", cols = green2purple,
           row.order = list(a=c("Monocytes","Macrophages","Mic.12","Mic.13","Other")), column.order = genes, scale.by = "col", cluster_rows=F, border=T, show_column_dend=F, cluster_columns=F)

pdf(file.path(panel.path, "4C.pdf"), width=10, height=3)
draw(hm[[1]], annotation_legend_list = hm[[2]], merge_legend=T)
while (!is.null(dev.list()))  dev.off()

rm(n.mic, exp, genes, hm)



# ----------------------------------------------------------------------------------------------------------------- #
#                                      Panel E - Markers Distributions - snRNAseq                                   #
# ----------------------------------------------------------------------------------------------------------------- #
# The following code produces scatter plots shown in panel 4e, as well as in Extended Data Figure 7d

cols <- c("Mic.12"="paleturquoise3", "Mic.13"="red3","Macrophages"="orchid3")

pdf(file.path(panel.path, "4E.pdf"), width=embed.width.small*3, height=embed.height.small*2)
ord <- list(c("TPRG1","CPM"), c("TPRG1","MRC1"), c("CPM","MRC1"))

snuc <- lapply(ord, function(pair) 
  validations$RNAscope$snuc.exp %>% 
    filter(state != "Other") %>%
    dplyr::select(state, g1=pair[[1]], g2=pair[[2]]) %>% 
    mutate(group = factor(paste(pair, collapse = "-")),
           state = factor(state, levels = rev(c("Mic.12","Mic.13","Macrophages"))))) %>% 
  do.call(rbind, .) %>% 
  rowwise() %>% 
  count(state, g1, g2, group) %>% 
  `[`(sample(rownames(.)),) %>%
  ggplot(., aes(g1,g2, color=state)) + 
  geom_jitter(alpha=.75) +
  scale_color_manual(values = cols) + 
  scale_x_continuous(expand = expansion(add=2)) + 
  scale_y_continuous(expand = expansion(add=2)) + 
  facet_wrap("group", scales = "free") + 
  labs(x=NULL, y=NULL, color=NULL) + 
  guides(colour = guide_legend(override.aes = list(size=3)))  +
  theme_classic() +
  theme(strip.background = element_blank())

df <- validations$RNAscope$df %>% 
  rowwise() %>% 
  mutate(state = c("Mic.12","Mic.13","Macrophages", "Other", "none")[which.max(c_across(c(Mic.12,Mic.13,Macrophages, Other, none)))]) %>%
  data.frame

scope <- lapply(ord, function(pair) df %>% dplyr::select(state, g1=pair[[1]], g2=pair[[2]]) %>% 
                  mutate(group = factor(paste(pair, collapse = "-")))) %>% 
  do.call(rbind, .) %>% 
  rowwise() %>% 
  count(state, g1, g2, group) %>% 
  mutate(state = factor(state, c("Mic.12","Mic.13","Macrophages"))) %>%
  `[`(sample(rownames(.)),) %>%
  ggplot(., aes(g1,g2, color=state)) + 
  scale_alpha_binned(range=c(.1, .5))+
  geom_jitter(alpha=.75) +
  scale_x_continuous(expand = expansion(add=.2)) + 
  scale_y_continuous(expand = expansion(add=.2)) + 
  scale_color_manual(values = cols, na.value = "lightgrey") + 
  guides(colour = guide_legend(override.aes = list(size=3)))  +
  facet_wrap("group", scales = "free") + 
  labs(x=NULL, y=NULL) + 
  theme_classic() +
  theme(strip.background = element_blank())

print(snuc / scope)
while (!is.null(dev.list()))  dev.off()
rm(snuc, scope, ord, df)



# ----------------------------------------------------------------------------------------------------------------- #
#                                        Panel F - Prevalence disease regression                                    #
# ----------------------------------------------------------------------------------------------------------------- #

pdf(file.path(panel.path, "4F.pdf"), width=embed.width.small, height=embed.height)
cols <- c("AD"="red3", "MCI"="chocolate2", "No AD"="#191970")
format <- paste(rep("t=%.2g, P=%.2g", 2), sep="*\", \"*")

merge(validations$RNAscope$predicted.proportions,
      validations$pTau$summarised, 
      by.x = c("row.names","batch","AD"), by.y = c("sample","batch","AD")) %>% 
  pivot_longer(cols = c(Mic.12,Mic.13), names_to = "state", values_to = "prop") %>% 
  ggplot(aes(pTau_tangles_plaques_median, prop)) + 
  geom_smooth(formula = "y~x", method="lm",color="black", alpha=.2) + 
  ggpmisc::stat_fit_tidy(aes(label=sprintf(format, after_stat(`x_stat`), after_stat(`x_p.value`)))) + 
  geom_point(aes(color=AD), size=2) + 
  scale_color_manual(values = cols) + 
  scale_y_continuous(expand = expansion(mult=.02), oob=scales::squish) + 
  scale_x_continuous(expand = expansion(mult=.01)) + 
  facet_wrap(~state, ncol=1, scales = "free") + 
  labs(x=NULL, y=NULL) + 
  theme_classic()  +
  theme(strip.background = element_blank(),
        legend.position = c(3,3))

rm(format, cols)
while (!is.null(dev.list()))  dev.off()



# ----------------------------------------------------------------------------------------------------------------- #
#                                        Panel G - associating morphology with state probability                    #
# ----------------------------------------------------------------------------------------------------------------- #

pdf(file.path(panel.path, "4G.pdf"), width=embed.width, height=embed.height)
split(validations$RNAscope$morpholoy.association, validations$RNAscope$morpholoy.association$trait) %>%
  lapply(., function(df) {
    v = max(abs(df$tstat))
    text = paste(round(df$tstat, 2), df$sig, sep = "\n")
    
    col.range <- green2purple(21)
    if(df[1,"trait"] == "eccentricity")
      col.range <- rev(col.range)
    
    Heatmap(df[,"tstat"],
            row_labels = df$state,
            cluster_rows = F,
            show_column_names = F,
            cluster_columns = F,
            width = unit(.8,"cm"),
            height = unit(8, "cm"),
            col = circlize::colorRamp2(seq(-v, v, length.out=21), col.range),
            cell_fun = function(j, i, x, y, w, h, col) grid.text(text[i], x,y),
            column_title = df[1,"trait"],
            row_names_side = "left",
            border=T,
            
            heatmap_legend_param = list(direction="horizontal", title="t stat"))
  }) %>% base::Reduce(`+`,.)
while (!is.null(dev.list()))  dev.off()



# ----------------------------------------------------------------------------------------------------------------- #
#                                             Panel H - PAM score for microglial states                             #
# ----------------------------------------------------------------------------------------------------------------- #
res <- validations$ROSMAP.PAM.Association
v   <- max(abs(res$t.value))

pdf(file.path(panel.path, "4H.pdf"), width=embed.width, height=embed.height)
plot_grid(
  Heatmap(res[1:4,"t.value"] %>% as.matrix(), 
          col = circlize::colorRamp2(seq(-v, v, length.out=7), green2purple(7)),
          row_labels = rownames(res)[1:4],
          cluster_rows = F,
          show_column_names = F,
          row_names_side = "left",
          border=T,
          width = unit(.8,"cm"),
          height = unit(8,"cm"),
          column_title="PAM Score",
          heatmap_legend_param = list(title="t stat"),
          cell_fun = function(j, i, x, y, w, h, col) grid.text(paste(round(res[i, "t.value"],2), res[i, "sig"], sep = "\n"), x, y)) %>% draw %>% grid.grabExpr(),
  
  Heatmap(res[5:6,"t.value"] %>% as.matrix(), 
          col = circlize::colorRamp2(seq(-v, v, length.out=7), green2purple(7)),
          cluster_rows = F,
          show_column_names = F,
          row_names_side = "left",
          border=T,
          width = unit(.8,"cm"),
          height = unit(4, "cm"),
          heatmap_legend_param = list(title="t stat"),
          cell_fun = function(j, i, x, y, w, h, col) grid.text(paste(round(res[i+4, "t.value"],2), res[i+4, "sig"], sep = "\n"), x, y)) %>% draw %>% grid.grabExpr(),
  align = "hv") %>% print(.)
while (!is.null(dev.list()))  dev.off()
rm(others, df, res, v, res.ctrl, controls)



# ----------------------------------------------------------------------------------------------------------------- #
#                                        Panel H - PAM scores with controls for confounders                         #
# ----------------------------------------------------------------------------------------------------------------- #
controls <- c("pmi", "Mic.12", "Mic.13")

pdf(file.path(panel.path, "4H.2.pdf"), width=embed.width, height=embed.height*2)
lapply(c("Mic.12","Mic.13"), function(state) {
  .df <- df %>% mutate(covariate = get(state))
  naive <- lm(pam ~ covariate, .df)
  
  # Controlling for confounding effects
  confounders <- lm(paste0("pam~", paste(setdiff(controls, state), collapse = "+")), .df)
  
  # Fitting pam to state after controlling for confounders
  fit <- lm(resid.pam ~ covariate, .df %>% mutate(resid.pam=residuals(confounders) + mean(.df$pam)))
  
  .df <- purrr::map2(list(naive, fit), c("direct", "controlled"), 
                     ~predict(.x, se=T) %>% `[`(c("fit","se.fit")) %>% as.data.frame %>% 
                       mutate(model=.y, state=state) %>% cbind(., .df[, c("pam","covariate")])) %>%
    do.call(rbind, .)

  ggplot(.df, aes(covariate, pam, group=model, fill=model, color=model)) + 
    geom_ribbon(aes(ymin=fit-se.fit, ymax=fit+se.fit), alpha=.1, color=NA) + 
    geom_line(aes(y=fit)) + 
    geom_point(data=. %>% filter(model == "direct"), color="black") + 
    labs(x=state, y="pam score")
}) %>% ggpubr::ggarrange(plotlist = ., ncol = 1, common.legend = T, legend="bottom")
while (!is.null(dev.list()))  dev.off()
rm(controls)




#####################################################################################################################
#                                       Supp Figure 4 - RNAscope Validations                                        #
#####################################################################################################################

# ----------------------------------------------------------------------------------------------------------------- #
#                              Panel A - Expression of key Mic.12/13 genes without collapsing states                #
# ----------------------------------------------------------------------------------------------------------------- #
genes = list(`Mic.12 genes` = c("CD163","HLA-DRA","FOXP1","PELI1","PELI2"),
             `Both`         = c("MSR1","PRKCE"),
             `Mic.13 genes` = c("PTPRG","PPARG","SPP1","TGFBR1","SMAD3","SCIN","WIPF3"),
             `AD risk genes`= c("ADAM10","TREM2","APOE","GPNMB"))
hm <- h5read(aggregated.data, "glia/microglia/gene.exp") %>% 
  filter(gene %in% unlist(genes)) %>%
  gheatmap(color.by = "mean.exp", size.by="pct.exp", cols = green2purple,
           row.order = list(a=atlas$microglia$state.order), column.order = list(a=unlist(genes)), scale.by = "col", cluster_rows=F, border=T, show_column_dend=F, cluster_columns=F)

pdf(file.path(panel.path, "s7A.pdf"), width=7.5, height=5)
draw(hm[[1]], annotation_legend_list = hm[[2]], merge_legend=T)
while (!is.null(dev.list()))  dev.off()

rm(genes, hm)



# ----------------------------------------------------------------------------------------------------------------- #
#                                     Panel B - Expression of validation markers                                    #
# ----------------------------------------------------------------------------------------------------------------- #
res <- h5read(aggregated.data, file.path(mapping[["microglia"]], "gene.exp")) %>% 
  filter(gene %in% c("CPM","TPRG1", "MRC1")) %>%
  mutate(pct.exp = 100*pct.exp) %>%
  gheatmap("state", "gene", "mean.exp", "pct.exp", 
           scale.by = "col",
           cols = green2purple,
           cluster_columns = F,
           row_order = atlas$microglia$state.order,
           column_order = c("CPM","TPRG1", "MRC1"),
           row_names_side="left",
           cluster_rows=F,
           row_title = " ",
           border=T,
           size.by.round=0)

pdf(file.path(panel.path, "s7B.pdf"), width=embed.width*.75, height=embed.height)
draw(res[[1]], merge_legends=T, annotation_legend_list = res[[2]])
while (!is.null(dev.list()))  dev.off()
rm(res)

